HC-984 Compilation errors (#70)
arduino/help-center-content: content/Software and Downloads/Compilation/If-your-sketch-doesnt-compile.md

---
title: "If your sketch doesn't compile"
---

Learn about compilation errors in your sketch and how to resolve them.

---

## Quick checks

* Make sure your error occurs during compilation by clicking ![Verify button](img/symbol_verify.png) **Verify** instead of ![Upload button](img/symbol_upload.png) **Upload**. This will compile the sketch without attempting to upload it. If your error only occurs when uploading, see [Errors when uploading a sketch](https://support.arduino.cc/hc/en-us/articles/4403365313810-Errors-when-uploading-a-sketch).

> * A successful compilation will always finish with this message:

Reviewer: ```suggestion
* A successful compilation will always finish with this message (the storage space and memory values might differ depending on the board used):
```
Just to make sure that the user don't assume that the message has to be _exactly_ the same for successful compilation

  ```
  Sketch uses 11604 bytes (4%) of program storage space. Maximum is 262144 bytes.
  Global variables use 2980 bytes (9%) of dynamic memory, leaving 29788 bytes for local variables. Maximum is 32768 bytes.
  ```

> * A sketch always needs to include a `setup()` and `loop()` function, even if they're not being used. You can use _File > Examples > 1.Basics > BareMinium_ as a template.

Reviewer: ```suggestion
* A sketch always needs to include a `setup()` and `loop()` function, even if they're not being used. You can use _File > Examples > 1.Basics > BareMinimum_ as a template.
```

2. Check the console output.
   * Compilation errors will often reference the row and column where the error was triggered, i.g. `Blink:29:1` (line 29, 1 column).
   * You may have to scroll or expand the window width to see the entire message.

3. Look for highlights in the editor:
   * In IDE 1.x, the **line** where the error occurred is highlighted in red.
   * In IDE 2.x, the **character** where the error occurred is underlined in red.
   * The line causing the error may be before the line where it was triggered. In the example above, line 28 is missing a semicolon (`;`), but the error will be triggered by the unexpected closing bracket (`}`) in line 29.

   <table class="img-captions">
     <tr>
        <td><img src="img/ide1-syntax-error-output-highlighted.png" alt="Compilation error in IDE 1.8.13, due to a missing semicolon."></td>
        <td><img src="img/ide2-syntax-error-output-highlighted.png" alt="Compilation error in IDE 2.0.0-3c6, due to a missing semicolon."></td>
     </tr>
     <tr>
        <td>Error output in IDE 1.x.</td>
        <td>Error output in IDE 2.x.</td>
     </tr>
   </table>

---

## Troubleshoot a specific error

> <table>

Reviewer: Missing links to anchors in some of the errors

Author: 6134b5e

<ul>
<li><a href="#required-functions">undefined reference to 'setup'</a></li>
<li><a href="#required-functions">undefined reference to 'loop'</a></li>
<li><a href="#required-functions">collect2: error: ld returned 1 exit status</a></li>
</ul>
<h3>Libraries</h3>
<ul>
<li><a href="#fatal-error-no-such-file-or-directory">fatal error: : No such file or directory</a></li>
</ul>
</td>
<td style="vertical-align: top;">
<h3>Syntax errors</h3>
<ul>
<li><a href="#">variable was not declared in this scope</a></li>
<li><a href="#">expected unqualified-id before numeric constant</a></li>
<li><a href="#">expected ',' or ';' before ':' token</a></li>
<li><a href="#"></a></li>
<li><a href="#"></a></li>
<li><a href="#"></a></li>
<li><a href="#"></a></li>
</ul>
</td>
</tr>
</table>

<!-- COMMON -->

<a id="missing-FQBN"></a>

### Compilation error: Missing FQBN (Fully Qualified Board Name)

When compiling or uploading code, this error will occur if no board is selected, or if the board core is not installed.

A board needs to be selected even if you're not uploading any code.

<a id="platform-not-installed"></a>

### Compilation error: Error: 2 UNKNOWN: platform not installed

> This error occurs when the required *board core* is not installed. Please ensure the core is installed using [this guide](https://www.arduino.cc/en/Tutorial/getting-started-with-ide-v2/ide-v2-board-manager#installing-a-core).

Reviewer: The link directs to the main IDE 2 tutorials, not to the installing a core guide

Author: 6134b5e

<!-- REQUIRED FUNCTIONS -->

<a id="required-functions"></a>

### undefined reference to 'setup' or 'loop'<br>collect2: error: ld returned 1 exit status

> Occurs when the `setup()` or `loop()` functions are missing. Your sketch _must_ include these functions, even if they're not being used. You can use _File > Examples > 1.Basics > BareMinium_ as a template. Note that function names are case-sensitive and that the compiler will something like `Setup()` (with a capital S) as a completely different function.

Reviewer: ```suggestion
Occurs when the `setup()` or `loop()` functions are missing. Your sketch _must_ include these functions, even if they're not being used. You can use _File > Examples > 1.Basics > BareMinimum_ as a template. Note that function names are case-sensitive and that the compiler will handle something like `Setup()` (with a capital S) as a completely different function.
```
